Zend/zend_enum.c

```diff
@@ -0,0 +1,338 @@
+/*
+   +----------------------------------------------------------------------+
+   | Zend Engine                                                          |
+   +----------------------------------------------------------------------+
+   | Copyright (c) Zend Technologies Ltd. (http://www.zend.com)           |
+   +----------------------------------------------------------------------+
+   | This source file is subject to version 2.00 of the Zend license,     |
+   | that is bundled with this package in the file LICENSE, and is        |
+   | available through the world-wide-web at the following url:           |
+   | http://www.zend.com/license/2_00.txt.                                |
+   | If you did not receive a copy of the Zend license and are unable to  |
+   | obtain it through the world-wide-web, please send a note to          |
+   | license@zend.com so we can mail you a copy immediately.              |
+   +----------------------------------------------------------------------+
+   | Authors: Ilija Tovilo <ilutov@php.net>                               |
+   +----------------------------------------------------------------------+
+*/
+
+#include "zend.h"
+#include "zend_API.h"
+#include "zend_compile.h"
+#include "zend_interfaces_arginfo.h"
+
+#define ZEND_ENUM_PROPERTY_ERROR() \
+	zend_throw_error(NULL, "Enum properties are immutable")
+
+#define ZEND_ENUM_DISALLOW_MAGIC_METHOD(propertyName, methodName) \
+	do { \
+		if (ce->propertyName) { \
+			zend_error_noreturn(E_COMPILE_ERROR, "Enum may not include %s", methodName); \
+		} \
+	} while (0);
+
+static zend_object_handlers enum_handlers;
+
+zend_object *zend_enum_new(zval *result, zend_class_entry *ce, zval *case_name_zv, zval *scalar_zv)
```

Comment:
Any reason to accept the case name as zval rather than zend_string?

Reply:
Mostly just that we're storing it in a property anyway which will require a zval. But to make the code clearer I can unwrap the string and then store it with `ZVAL_STR_COPY`.